pixeltable/catalog/table.py

```diff
@@ -1550,3 +1552,138 @@ def __dir__(self) -> list[str]:
+        """Return a string reporting the difference between two schema versions"""
+        assert new_md is not None
+        if old_md is None:
+            return 'Initial Version'
+        if old_md == new_md:
+            return ''
+        added = {k: v['name'] for k, v in new_md.items() if k not in old_md}
+        changed = {
+            k: f'{old_md[k]["name"]} to {v["name"]}'
+            for k, v in new_md.items()
+            if k in old_md and old_md[k]['name'] != v['name']
+        }
+        deleted = {k: v['name'] for k, v in old_md.items() if k not in new_md}
+        if len(added) == 0 and len(changed) == 0 and len(deleted) == 0:
+            return ''
+        # Format the result
+        t = []
+        if len(added) > 0:
+            t.append('Added: ' + ', '.join(added.values()))
+        if len(changed) > 0:
+            t.append('Changed: ' + ', '.join(changed.values()))
+        if len(deleted) > 0:
+            t.append('Deleted: ' + ', '.join(deleted.values()))
+        r = ', '.join(t)
+        return r
+
+    @classmethod
+    def _create_md_change_dict(cls, md_list: list[tuple[int, dict[str, Any]]]) -> dict[int, str]:
+        """Return a dictionary of schema changes by version
+        Args:
+            md_list: a list of tuples, each containing a version number and a metadata dictionary.
+        """
+        r: dict[int, str] = {}
+        if md_list is None or len(md_list) == 0:
+            return r
+        md_list.sort()
+
+        v0 = md_list[0][0]
+        if v0 == 0:
+            ofld = None
+            ver_old = -1
+            start = 0
+        else:
+            ofld = md_list[0][1]
+            ver_old = v0
+            start = 1
+
+        for row in md_list[start:]:
+            ver = row[0]
+            if ver == ver_old:
+                continue
+            assert ver > ver_old
+            nfld = row[1]
+            tf = cls._diff_md(ofld, nfld)
+            ofld = nfld
+            if tf != '':
+                r[ver] = tf
+        return r
+
+    def history(self, max_versions: int = 1_000_000_000) -> pixeltable.dataframe.DataFrameResultSet:
+        """Returns rows of information about the versions of this table, most recent first.
+
+        Args:
+            max_versions: a limit to the number of versions listed
+
+        Examples:
+            Report history:
+
+            >>> tbl.history()
+
+            Report only the most recent 5 changes to the table:
+
+            >>> tbl.history(max_versions=5)
+
+        Returns:
+            A list of rows of data.
+        """
+        if not isinstance(max_versions, int) or max_versions < 1:
+            raise excs.Error(f'Invalid value for max_versions: {max_versions}')
+
+        with Env.get().begin_xact():
+            # tbl_id = self._tbl_version_path.tbl_id
+            tbl_id = self._id
+            if tbl_id is None:
+                raise excs.Error('Cannot get history of a table that has been dropped')
+            q = (
```

Comment:
Table should not be running queries against the stored metadata itself. It's only a user-facing interface class. All metadata retrieval is handled by Catalog.

See Catalog.get_table() for the expected pattern (retry_loop(), etc.).